### PR TITLE
[WIP] INSERT ... ON CONFLICT support for postgres

### PR DESCRIPTION
**Pull request by jordansexton**
This is based in part on [this gist](https://gist.github.com/hayeah/1c8d642df5cfeabc2a5b) by @hayeah, and in part on the existing update method implementation.

It would affect #1633, #1351, #1121, and #54. I wrote preliminary tests for it, which I'm using the CI to run, but I'd like feedback on the them (the test files are rather hard to parse).

I don't think any consensus has been reached on a more generalized "upsert" feature, but I think supporting this in a database-specific way is reasonable and makes the API easy to understand.

Asking @elhigu (because you had input on #54), do you think a Postgres-specific approach is acceptable, or would this need to be expanded to support all clients that have some kind of `MERGE`/`ON DUPLICATE KEY UPDATE` functionality?

### CODE REVIEW

**Comment by jordansexton** [comment]
Tests appear to be broken on:

```
Selects
      ✓ runs with no conditions
      ✓ returns an array of a single column with `pluck`
      ✓ can pluck a qualified column name, #1619
      ✓ starts selecting at offset
      ✓ returns a single entry with first
      1) allows you to stream
```

Not sure why, nothing relating to streaming would have changed.

**Comment by elhigu** [comment]
I'm running tests again to see if it was some random problem (didn't look like random, since all node versions failed though).

**Comment by jordansexton** [comment]
Thanks! I haven't been able to get the test environment running on my local machine yet, so I haven't been able to debug the failure.

File: src/query/builder.js [hunk 1]
```diff
@@ -853,6 +853,15 @@ assign(Builder.prototype, {
     return this;
   },
 
+  // Insert on conflict do nothing/update
+  onConflict(columns, updates) {
```
Comment (elhigu) [inline src/query/builder.js]: I think this should also allow passing raw() so that people can also write their own code, what to do in case of conflict (sqlite has quite a different selection than nothin / update set https://sqlite.org/lang_conflict.html).

File: test/integration/builder/inserts.js [hunk 2]
```diff
@@ -648,6 +648,122 @@ module.exports = function(knex) {
...
+          .testSql(function(tester) {
+            tester(
+              'postgresql',
+              'insert into "accounts" ("email", "first_name", "last_name") values (?, ?, ?) on conflict ("email") do update set "first_name" = \'Testing\', "last_name" = \'Users\' returning "id"',
+              ['test8@example.com','Test','User'],
+              ['9']
+            );
+          });
+      }
+    });
```
Comment (elhigu) [inline test/integration/builder/inserts.js]: In addition to unit tests which are just generating queries,  there should be real integration test, which actually tries to insert stuff to db and checks that in conflict situation there was correct data added to DB.

**Comment by elhigu** [comment]
@jordansexton I just gave this an initial review how the API and design looks compared to other DBs which may support upsert + basic notes of minimum testing/documentation requirements. Thanks for the pull request though! I'm really happy to see that someone finally started to work on this 😊

**Comment by elhigu** [comment]
> Thanks! I haven't been able to get the test environment running on my local machine yet, so I haven't been able to debug the failure.
--

I will figure out why those tests were failing, I don't believe it has anything to do with this PR.

**Comment by jordansexton** [comment]
Thanks! 👍 All your proposals sound good to me, I'll work on this. Do you have any tips or documentation for how to write unit and integration tests for knex? I just copied the`tester(...)` blocks from elsewhere, but the indirection of the system feels opaque.

**Comment by jordansexton** [comment]
And are you comfortable with calling it `onConflict`? MySQL's version of the same is pretty wordy...

**Comment by elhigu** [comment]
> And are you comfortable with calling it onConflict? MySQL's version of the same is pretty wordy...

Something more generic, which is not exactly the wording of postgresql nor mysql would be better to keep API less opinionated how it is implemented... maybe `onUpsert()` ? sounds a bit weird since end result is `.insert().onUpsert()`, but I can live with that. As far as I know none of the BDs is actually calling their upsert syntax as `upsert` :)

**Comment by elhigu** [comment]
For writing real integration tests, this was pretty clean way to do it (I just browsed back commits until I found this) https://github.com/tgriesser/knex/commit/5fc456cb693a12d789167bb1eeac006edf5f9f17

**Comment by jordansexton** [comment]
Here's a brief rundown of the differing APIs for the PostgreSQL, MySQL, and SQLite:

- Postgres's `ON CONFLICT DO UPDATE` **requires** passing column/index names *if* updates are provided.

> For `ON CONFLICT DO NOTHING`, it is optional to specify a conflict_target; when omitted, conflicts with all usable constraints (and unique indexes) are handled. For `ON CONFLICT DO UPDATE`, a `conflict_target` must be provided.

- MySQL's `ON DUPLICATE KEY UPDATE` doesn't support passing columns at all. The alternative syntax is for doing nothing is `INSERT IGNORE`.

- SQLite's `INSERT OR IGNORE INTO` / `INSERT OR REPLACE INTO` (or just `REPLACE INTO`) also doesn't support passing conflicting columns or indexes.

Arguments against `onUpsert`:

- No supported RDBMS actually calls the operation "upsert", nor is it standard SQL. The SQL standard uses `MERGE` (this is what is implemented by Oracle and MS SQL).
- `insert(...).onUpsert(...)` not only looks confusing, but also ultimately confuses what's happening; makes very little sense in the `IGNORE`/`DO NOTHING` case.
- It sounds like registering an event listener.

Arguments against `onConflict`

- Only Postgres calls it that (technically, SQLite has `ON CONFLICT`, but you just use the shorthand `INSERT OR ...` / `REPLACE INTO` above instead.
- `insert(...).onConflict(...)` itself reads well, but then the parameters passed to `onConflict` aren't consistent between RDBMSs, and even for Postgres, aren't self-documenting. You just have to know what the code you're reading is supposed to do.
- It also sounds like registering an event listener.

In light of the above, I think using a chaining API that better indicates what you want to accomplish might be better.

---

Possible alternatives:

#### 1. `ignore` and `merge`
```javascript
insert({ ...inserts }).ignore([columns]?)
insert({ ...inserts }).merge({ ...updates }, [columns]?)
```
- **Pros**
  - Readable, explicit, no magic.
  - Columns optional for Postgres (or any other DB).
- **Cons**
  - While `MERGE` is in the SQL standard, this _type_ of operation is often referred to as an upsert by devs.
      - `merge` in a chaining API context passing around object literals is questionable, because it most often implies merging objects.
- **Pro or Con**
  - Two methods instead of one. I might consider this a pro, as I tend to prefer an explicit API, even if it means the API surface is larger as a result. Two methods also means it's more easily composed by application code that... builds the query... builder.

#### 2. `ignore` and `replace`
```javascript
insert({ ...inserts }).ignore([columns]?)
insert({ ...inserts }).replace({ ...updates }, [columns]?)
```
- **Pros**
  - Same as above, but `replace` is more contextually meaningful than `merge`.
- **Cons**
  - `REPLACE` itself isn't standard SQL, and is only used by SQLite.

#### 3. `upsert` or `onUpsert`
```javascript
insert({ ...inserts }).upsert({ ...updates }, [columns]?)
insert({ ...inserts }).onUpsert({ ...updates }, [columns]?)
```
- **Pros**
  - Uses `upsert`, which despite not being used by any supported RDBMS, is commonly used by devs.
- **Cons**
  - `insert` followed by `upsert` is pretty weird.
  - `insert` followed by `onUpsert` confuses what is actually happening: you're doing doing something "on upsert" (which itself isn't a thing), but you're doing "an upsert" upon the insert failing.
  - This is what it looks like when you want to `IGNORE`: `insert({ foo: 'bar' }).onUpsert()` or worse, in Postgres, `insert({ foo: 'bar' }).onUpsert(['foo'])`. This is confusing, and I think it's for the reason that when you call `upsert` or `onUpsert` (with or without arguments), it makes it look like something is going to happen, but what you actually want is for nothing to happen.
  - The semantics of `upsert` or `onUpsert` only make sense in the positive case; in the negative case, they look magical (and then we have to document that if the `{ ...updates }` hash isn't passed, you have to pass the columns as the first arg (or we check if it's an array). Ultimately, the semantics of the API don't follow directly from its intentions.

#### 4. `upsert`
```javascript
upsert({ ...inserts }, { ...updates }, [columns]?)
```
- **Pros**
  - One method, one new operation.
  - Semantically makes sense; devs often refer to this operation atomically as such.
- **Cons**
  - A single method hides what's actually going on (in `insert` or an `update`), and doesn't play as nicely with application code that wants to build a query in pieces.
  - Requires replacing existing `inserts` in your code rather than adding calls onto them.
  - With lots of columns, or with raw values, this gets kind of ugly, especially when you have raw values as part of your update (common for incrementing columns on conflict, aggregation, pulling values from other columns, etc.)
  - As with `onUpsert`, magical when you want to `IGNORE`. The API is `upsert({ ...inserts })`, which will do nothing, or in Postgres, `upsert({ ...inserts }, [columns])`.

---

My personal preference is for having two methods, one to ignore/do nothing, and one to perform an update. You would never actually use both, but because you are doing one of two different things, you should invoke methods that semantically match your intent.

Thoughts?

**Comment by jordansexton** [comment]
Hmm, supporting this in SQLite with a simple implementation is dubious.

> When a UNIQUE or PRIMARY KEY constraint violation occurs, the REPLACE algorithm deletes pre-existing rows that are causing the constraint violation prior to inserting or updating the current row and the command continues executing normally. If a NOT NULL constraint violation occurs, the REPLACE conflict resolution replaces the NULL value with the default value for that column, or if the column has no default value, then the ABORT algorithm is used. If a CHECK constraint violation occurs, the REPLACE conflict resolution algorithm always works like ABORT.
~ https://sqlite.org/lang_conflict.html

I understand this to mean that `insert({ ...inserts }).upsert/onUpsert/merge/replace({ ...updates })` is not going to patch with just the updates provided, but is instead going to wipe all the columns not provided to their defaults. This is a pretty big footgun and is not what I would expect as a user of the library.

https://github.com/tgriesser/knex/issues/54#issuecomment-289217529 has some description of it.

This (the fact that it actually deletes the row then inserts a new one, rather than updating the row) suggests to me that a `INSERT OR REPLACE` aka `REPLACE INTO` is a fundamentally different operation, and perhaps should not be supported in the same way.

**Comment by jordansexton** [comment]
#### Problems with SQLite

I think [this SO answer](https://stackoverflow.com/a/418988) is probably the most sound approach, but outside the scope of this feature. I think it's probably best to not support this on SQLite since it's not a safe operation, and the fix options ([1](https://stackoverflow.com/a/7511635), [2](https://stackoverflow.com/a/4330694)) for it would also be very complex/opinionated, and require Knex to know how to select the right row, since SQLite doesn't do this automatically.

The alternative is supporting this in SQLite with just a documentation warning, but that will undoubtedly result in people's data getting deleted if they get their query wrong. I think having the feature poorly or partially implemented is worse than not having it, and it will lead SQLite users to believe that it's equivalent.

#### Problems with `MERGE`

Oracle and MS SQL support the SQL standard `MERGE` operation (or some variant thereof), which takes the form:

```sql
MERGE INTO table_name AS target
USING table_name AS source
    ON target.some_column = source.some_column
    ...
WHEN MATCHED THEN 
    UPDATE SET ...
WHEN NOT MATCHED BY TARGET THEN
    INSERT (...) VALUES (...);
```

The issue here is that the requisite table aliases and condition columns can't be known with any of the API proposals we have so far. We'd have to do something like

```javascript
insert({ ...inserts }).merge({ ...updates }, [conditions]))
```

where `conditions` is maybe one or more of:
- a callback that allows building the merge with a DSL, similar to a join callback (doable, complicated)
- a raw query (problematic because you still need to define the table aliases)
- an array of identifiers to match on (same issue as a raw query)

Alternatively, `merge` needs a `targetAlias` argument as well as `conditions`, which would make raw queries or identifiers doable, but still complicated. There's also the issue of what happens when you want to effectively `IGNORE`/`DO NOTHING`

Ultimately, the API for a SQL standard MERGE needs additional params that can't be provided by this API.

#### Conclusion

I don't think it's useful to implement this in a wholly database-agnostic fashion. The syntax and semantics for MySQL and Postgres are very close; the only difference we care about is the requirement to specify the column(s)/index(es) in Postgres if `DO UPDATE`-ing.

For `MERGE` and `REPLACE INTO`, the functionality is quite different. I think it would ultimately be bad to shoehorn them all into the same API.

**Comment by elhigu** [comment]
Thanks for investigating this throughly!

> Problems with SQLite

I agree. If the implementation would be somewhat hacky better not to support that platform and allow people to decide how to deal with it. 

> My personal preference is for having two methods, one to ignore/do nothing, and one to perform an update. You would never actually use both, but because you are doing one of two different things, you should invoke methods that semantically match your intent.

I think this one is nicer instead of having single method. 

> I don't think it's useful to implement this in a wholly database-agnostic fashion. The syntax and semantics for MySQL and Postgres are very close; the only difference we care about is the requirement to specify the column(s)/index(es) in Postgres if DO UPDATE-ing.

> For MERGE and REPLACE INTO, the functionality is quite different. I think it would ultimately be bad to shoehorn them all into the same API.

I don't believe in it either, but many knex users are actually using knex for abstracting DBs as much as they can, so in that sense it would be convenient. 

It will naturally mean that the DB agnostic way will not be as powerful as separate APIs for each DB. I would like to go with agnostic udpate / ignore way at first and then people who doesn't want to use raw() for having DB specific functionality, they could add also DB specific APIs for certain DBs.

Could simple update on conflict case work somehow like this:
```
knex('blah').insert({id: 1, comment: 'the comment' }).replace({ comment: 'was replaced' }, ['id'])
```

```
MERGE INTO  blah AS target
USING blah AS source
    ON target.id = source.id
WHEN MATCHED THEN 
    UPDATE SET comment = 'was replaced'
WHEN NOT MATCHED BY TARGET THEN
    INSERT (...) VALUES (...);
```

Does this sound viable way to do it with merge? 

You already understand this problem a lot better than I do and all the stuff you've written sounds legit so I trust your judgement on this.  Anyways I would like to emphasis that more generic and less powerful method would probably serve bigger portion of the knex userbase since update/ignore on conflict seems to be the thing that most of the people are looking for.

**Comment by jordansexton** [comment]
@elhigu Sounds good. I will leave SQLite out for now and implement and test on Postgres and Mysql with two methods.

The problem I see with implementing `MERGE` for MS SQL/Oracle is that if we hardcode `source` and `target`, the developer who is doing the operation can't name the aliases, can't have any conflicting aliases for the query, and has to know those are the aliases to refer to them in raw updates.

If this is acceptable to you, I can work on that as well. If not, I will just get PG and MySQL working, and we can revisit this.

**Comment by elhigu** [comment]
@jordansexton good point about the aliases. 

We can hardcode them to something that probably won't collide with users own table names and it can be improved later on to allow overriding default alias names. `mergeSource` and `mergeTarget` could be unique enough?

Its just important to add default names to docuementation with an example usage.

**Comment by jordansexton** [comment]
Okay. I'll finish the implementation for `DO NOTHING/UPDATE` on Postgres and `INSERT IGNORE/ON DUPLICATE KEY UPDATE` on MySQL in this PR, probably using two methods:
```javascript
insert({ ...inserts }).ignore([columns]?)
insert({ ...inserts }).merge({ ...updates }, [columns]?)
```

My reasoning is:

1. `replace` implies full replacement, whereas `merge` implies partial replacement (essentially, the semantics for a generic "object merge" are equivalent to what actually happens in the operation)
2. Only SQLite uses replace. Since this PR won't support SQLite, and replace is a semantically different operation, it's probably better to have a different `replace` method (if supported at all).
3. The SQL standard uses `MERGE`. In the event that other RDBMS's converge upon the standard, `merge` is likely more future-proof than an ad-hoc alternative.
4. A followup PR to support MS SQL and Oracle will be able to use the same method and just work. I don't use either, but I'll take a stab at supporting both when this is settled. The second `columns` argument could be used to provide aliases for those instead.

**Comment by Druotic** [comment]
@jordansexton thanks for your work on this! It'll be great being able to avoid having to use raw queries to get upsert behavior. @elhigu What's left to get an initial iteration of this merged? I see comments about an end-to-end integration test as well as being more db agnostic/supporting additional DBs with this behavior.

**Comment by elhigu** [comment]
@Druotic This initial PR will not be merged, but it needs pretty much complete rework to change the implementation to match APIs described in https://github.com/tgriesser/knex/pull/2197#issuecomment-325934623 and tests  and documentation accordingly.

@jordansexton indeed already did awesome work figuring out good database agnostic APIs to support upserty behavior. 

This is pretty much the last thing that I would like to see finished before releasing knex 0.14

**Comment by jordansexton** [comment]
@elhigu @Druotic Thanks for the ping. I've been pretty busy and spaced on this, but I'll see what I can do about it this weekend. And yes, I plan to implement the API described @ https://github.com/tgriesser/knex/pull/2197#issuecomment-325934623. Any help with the testing is very much appreciated, as I found the test system very inscrutable.

**Comment by austinkelleher** [comment]
I really need this feature. Can't wait to see this get merged. Thanks @jordansexton!

**Comment by elhigu** [comment]
@jordansexton I'm starting to work on release 0.14, if this misses it I can make 0.14.1 when this is ready.

**Comment by nihiluis** [comment]
whats the status? its version 0.14.2 now

**Comment by elhigu** [comment]
@nihiluis Feature is designed, but no one have had time to implement it yet. As soon as its done I will do an additional patch release to get it out (except that I'll be in vacations couple of week before christmas).

**Comment by tsofist** [comment]
Still.. no? 🔍

**Comment by ivands** [comment]
I would really like to see this feature included. Any updates?

**Comment by elhigu** [comment]
Just to keep important info in the end of this thread:

> This initial PR will not be merged, but it needs pretty much complete rework to change the implementation to match APIs described in https://github.com/tgriesser/knex/pull/2197#issuecomment-325934623 and tests  and documentation accordingly.
>
> @jordansexton indeed already did awesome work figuring out good database agnostic APIs to support upserty behavior. 

So if someone is up to implement that API, please do open new PR with implementation! Locking conversation to prevent statusupdateflood, but not closing since this is the only place where DB agnostic upsert API has been discussed throughly.

**Comment by tgriesser** [comment]
Unlocking this as I'd like to take a look at getting this into knex soon. Sorry for the delay everyone

**Comment by obataku** [comment]
how long til it's merged?

**Comment by elhigu** [comment]
@obtaku never, this PR has only plan how the feature should be implemented. There is no code matching that plan... maybe ill just close this and collect all the relevant info to feature request to prevent further confusion.

**Comment by elhigu** [comment]
Created feature request
